python/cudf_polars/cudf_polars/containers/datatype.py

```diff
@@ -0,0 +1,123 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024-2025, NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+
+"""A datatype, preserving polars metadata."""
+
+from __future__ import annotations
+
+from functools import cache
+
+from typing_extensions import assert_never
+
+import polars as pl
+
+import pylibcudf as plc
+
+__all__ = ["DataType"]
+
+
+@cache
+def _from_polars(dtype: pl.DataType) -> plc.DataType:
+    """
+    Convert a polars datatype to a pylibcudf one.
+
+    Parameters
+    ----------
+    dtype
+        Polars dtype to convert
+
+    Returns
+    -------
+    Matching pylibcudf DataType object.
+
+    Raises
+    ------
+    NotImplementedError
+        For unsupported conversions.
+    """
+    if isinstance(dtype, pl.Boolean):
+        return plc.DataType(plc.TypeId.BOOL8)
+    elif isinstance(dtype, pl.Int8):
+        return plc.DataType(plc.TypeId.INT8)
+    elif isinstance(dtype, pl.Int16):
+        return plc.DataType(plc.TypeId.INT16)
+    elif isinstance(dtype, pl.Int32):
+        return plc.DataType(plc.TypeId.INT32)
+    elif isinstance(dtype, pl.Int64):
+        return plc.DataType(plc.TypeId.INT64)
+    if isinstance(dtype, pl.UInt8):
+        return plc.DataType(plc.TypeId.UINT8)
+    elif isinstance(dtype, pl.UInt16):
+        return plc.DataType(plc.TypeId.UINT16)
+    elif isinstance(dtype, pl.UInt32):
+        return plc.DataType(plc.TypeId.UINT32)
+    elif isinstance(dtype, pl.UInt64):
+        return plc.DataType(plc.TypeId.UINT64)
+    elif isinstance(dtype, pl.Float32):
+        return plc.DataType(plc.TypeId.FLOAT32)
+    elif isinstance(dtype, pl.Float64):
```

Comment:
@mroeschke - Should these be instance checks like this? I'm debugging a multi-GPU Polars hang, and the cause seems to be that `_from_polars` is being handed a `dtype` argument of `pl.Float64`, and this instance check is returning `False`.

Reply:
This was moved from `python/cudf_polars/cudf_polars/utils/dtypes.py` (in this diff).

To confirm, you're seeing `pl.Float64` the class and not an instance? I'm surprised you're not seeing the `NotImplementedError` being raise anywhere below

Reply:
>To confirm, you're seeing pl.Float64 the class and not an instance? I'm surprised you're not seeing the NotImplementedError being raise anywhere below

Sorry - Yes, I'm hitting the `NotImplementedError` on line 94. Not entirely sure why this leads to a hang in Dask yet.